File: python/39.py

def num_solution(p):
    s = 0
    for a in range(1, p//2):
        b = (p**2 - 2*a*p) / (2*p - 2*a)
        if int(b) == b:
            s += 1
    return s // 2


assert num_solution(120) == 3

M, p = 0, 0

for i in range(1, 1001):
    s = num_solution(i)
    if s > M:
        M, p = s, i

print(p)

File: cpp/39.cpp
#include <iostream>
#include <math.h>

using namespace std;


int num_solution(int p){
    int s {};
    for (int a = 1; a < p / 2; a++)
    {
        float b = (p * p - 2 * a * p) / float(2 * p - 2 * a);

        if (ceil(b) == b)
            s++;
    }
    
    return s / 2;
}


int main() {
    int M {}, p {}, s{};

    for (int i = 1; i < 1001; i++)
    {
        s = num_solution(i);
        if (s > M){
            M = s;
            p = i;
        }
    }
    
    cout << p << endl;

    return 0;
}